File: Comunication/client_socket_topic_PC.py
#! /usr/bin/env python

import socket
import rospy
import actionlib

import action\PWMAction

## Socket address
HOST = '192.168.0.42'
PORT = 8888


pwm_values = [0,0,0,0,0,0,0,0,0,0,0,0]
distance_U = 0
accel_X = 0
accel_Y = 0
accel_Z = 0
gyro_X = 0
gyro_Y = 0
gyro_Z = 0

"""
ArduinoController runs the services related with the arduino:
	/arduino/ultrasound
	/arduino/accel
	/arduino/pwm
"""

def pwm_action_cb(goal):
	rospy.loginfo('/arduino/pwm service callback executed')
	PWM_action.set_succeeded()
	#callback of pwm_action

def main():
	s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
	rospy.loginfo('Socket created')

	s.connect((HOST,PORT))
	rospy.loginfo('Socket connected')


	PWM_action = actionlib.SimpleActionServer('/arduino/pwm', PWMAction, pwm_service_cb, False)
	PWM_action.start()
	rospy.loginfo('Action /arduino/pwm launched')

	## Main loop
	while(1):	

		s.send(100)
		distance_U = int(s.recv(1024))

		s.send(101)
		accel_X = int(s.recv(1024))
		s.send(102)
		accel_Y = int(s.recv(1024))
		s.send(103)
		accel_Z = int(s.recv(1024))

		s.send(104)
		gyro_X = int(s.recv(1024))
		s.send(105)
		gyro_Y = int(s.recv(1024))
		s.send(106)
		gyro_Z = int(s.recv(1024))

		for i in range(1,13):
			s.send(i)
			s.send(pwm_values[i])

		rospy.loginfo('Values sent')




if __name__ == '__main__':
	rospy.init_node('client_socket_topic')
	rospy.loginfo("Initializing client_socket_topic")
	main()

File: catkin_ws/src/communication_pkg/src/client_socket_topic_PC.cpp

#include <communication_pkg/PWMAction.h>
#include <actionlib/server/simple_action_server.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <sstream>


// Socket address
#define HOST "192.168.0.42"
#define PORT 8888

char pwm_values[12] = {0,0,0,0,0,0,0,0,0,0,0,0};
int distance_U = 0;
int accel_X = 0;
int accel_Y = 0;
int accel_Z = 0;
int gyro_X = 0;
int gyro_Y = 0;
int gyro_Z = 0;

typedef actionlib::SimpleActionServer<communication_pkg::PWMAction> Server;

void error(const char *msg)
{
    perror(msg);
    exit(0);
}

void execute(const communication_pkg::PWMGoalConstPtr& goal, Server* as)
{
	char aux[100];
	as->setSucceeded();
	
	//sprintf(aux,"goal 1:%d, goal 12:%d\n",(*goal).pwm[0],(*goal).pwm[11]);
	sprintf(aux,"/arduino/pwm service callback executed");
	write(1,aux,strlen(aux));
}

int main(int argc, char** argv)
{
	int s,n;
	struct sockaddr_in addr;
	char buff[256];

	ros::init(argc, argv, "socket_client_topic_server");
	printf("Initializing socket_client_topic_server\n");
	ros::NodeHandle h;


	s = socket(AF_INET, SOCK_STREAM, 0);
	if (s==-1)
		perror("Creat socket:");
	else
		printf("Socket created\n");
	
    	addr.sin_family = AF_INET;
	addr.sin_port = htons(PORT);
	inet_aton(HOST, &addr.sin_addr);

	if (connect(s,(struct sockaddr *) &addr,sizeof(addr))<0)
		perror("Error connecting socket:");
	else
		printf("Socket connected\n");


	Server server(h, "/arduino/pwm", boost::bind(&execute, _1, &server), false);
	server.start();
	printf("Action /arduino/pwm launched\n");
	

	// Main loop
	while(1){	
	// Ultrasound
		sprintf(buff,"100");
		write(s,buff,strlen(buff));
		//printf("writen %s in socket\n",buff);	

		memset(buff,0,strlen(buff));
		n = read(s,buff,255);
		distance_U = atoi(buff);
		if (n == 0)		printf("nothing read\n");
		else if (n<0)		perror("Read Error:");
		//else			
		printf("distance_U = %d\n",distance_U);

	// Accelerometer
		sprintf(buff,"101");		write(s,buff,strlen(buff));
		memset(buff,0,strlen(buff));
		n = read(s,buff,255);		accel_X = atoi(buff);
		printf("accel_X = %d\n",accel_X);

		sprintf(buff,"102");		write(s,buff,strlen(buff));
		memset(buff,0,strlen(buff));
		n = read(s,buff,255);		accel_Y = atoi(buff);
		printf("accel_Y = %d\n",accel_Y);

		sprintf(buff,"103");		write(s,buff,strlen(buff));
		memset(buff,0,strlen(buff));
		n = read(s,buff,255);		accel_Z = atoi(buff);
		printf("accel_Z = %d\n",accel_Z);

	// Gyroscope
		sprintf(buff,"104");		write(s,buff,strlen(buff));
		memset(buff,0,strlen(buff));
		n = read(s,buff,255);		gyro_X = atoi(buff);
		printf("gyro_X = %d\n",gyro_X);

		sprintf(buff,"105");		write(s,buff,strlen(buff));
		memset(buff,0,strlen(buff));
		n = read(s,buff,255);		gyro_Y = atoi(buff);
		printf("gyro_Y = %d\n",gyro_Y);

		sprintf(buff,"106");		write(s,buff,strlen(buff));
		memset(buff,0,strlen(buff));
		n = read(s,buff,255);		gyro_Z = atoi(buff);
		printf("gyro_Z = %d\n",gyro_Z);

	// PWM
		/*for (int i=0;i<12;i++){
			std::stringstream stream;	
			stream << i;
			write(s,stream.str(),stream.gcount());	//send the value of 'i'
			memset(buff,0,strlen(buff));	read(s,buff,255);	memset(buff,0,strlen(buff));	//wait for acknowledge
			stream << pwm_values[i]
			write(s,stream.str(),stream.gcount());	//send pwm of servo 'i'
			printf("pwm_value[%d] = %d\n",i,pwm_values[i]);
		}*/

		printf("Values sent\n");

		ros::spinOnce();
		ros::Duration(2).sleep();
	}
	return 0;
}
